### PR TITLE
synapse: range(...) arguments type check

### PR DESCRIPTION
**Pull request by dsprenkels**
Issue #781.

This PR adds a check and tests to ensure that during synapse generation, arguments to `range` are always `int`s. If not, it will raise with a **`TypeError`**.

### CODE REVIEW

File: brian2/synapses/synapses.py [hunk 1]
```diff
@@ -1570,3 +1575,17 @@ def _add_synapses_generator(self, j, n, skip_if_invalid=False, namespace=None, l
                                         check_units=False,
                                         run_namespace=namespace)
         codeobj()
+
+    def _check_synapse_parse(self, parsed, namespace):
```
Comment (mstimberg) [inline brian2/synapses/synapses.py]: Maybe we can find a clearer name for this function? `_check_generator_arguments` or something along these lines?

File: brian2/tests/test_synapses.py [hunk 2]
```diff
@@ -2190,6 +2191,22 @@ def test_synapses_refractory_rand():
...
     assert_equal(target.v[5:], 0)
 
+
+@attr('codegen-independent')
+def test_range_noint_error():
+    # arguments to `range` should only be integers (issue #781)
+    G = NeuronGroup(42, 'v: 1', threshold='False')
+    S = Synapses(G, G, 'w:1', 'v+=w')
+    msg = 'The "{}" argument of the range function was .+, but it needs to be an integer.'
+    assert_raises_regex(TypeError, msg.format('high'), lambda: S.connect(j='k for k in range(42.0)'))
```
Comment (mstimberg) [inline brian2/tests/test_synapses.py]: Good test, I wonder if it is not a bit too specific by matching the exact text of the error message. On the other hand, it should be pretty obvious what the issue is, if we ever to decide to change the text. So, ok, let's leave it that way.

File: brian2/synapses/synapses.py [hunk 3]
```diff
@@ -1570,3 +1575,17 @@ def _add_synapses_generator(self, j, n, skip_if_invalid=False, namespace=None, l
...
+        if parsed['iterator_func'] == 'range':
+            # We expect all arguments of the range function to be integers
+            for argname, arg in parsed['iterator_kwds'].items():
+                identifiers = get_identifiers(arg)
+                variables = self.resolve_all(identifiers, run_namespace=namespace,
+                                             user_identifiers=identifiers)
+                annotated = brian_ast(arg, variables)
+                print(annotated, annotated.dtype)
```
Comment (mstimberg) [inline brian2/synapses/synapses.py]: Please remove the print statement.

File: brian2/tests/test_synapses.py [hunk 4]
```diff
@@ -2264,4 +2281,6 @@ def test_synapses_refractory_rand():
     test_synapse_generator_random_with_condition()
     test_synapses_refractory()
     test_synapses_refractory_rand()
+    test_synapses_refractory()
```
Comment (mstimberg) [inline brian2/tests/test_synapses.py]: This should call your new test, instead of calling the refractory tests twice.

Comment (dsprenkels) [inline brian2/tests/test_synapses.py]: This must have gone wrong during merge conflict resolving. :+1:

File: brian2/synapses/synapses.py (hunk 1 above)
Comment (dsprenkels) [inline brian2/synapses/synapses.py]: I'll change it to `_check_parsed_synapse_generator`. But I will add a suitable doc string to clarify the function's purpose.

Comment (dsprenkels) [inline brian2/synapses/synapses.py]: The idea would be to extend this function if any other type-checking is needed in the future. Maybe this is a bit too elaborate?

Comment (mstimberg) [inline brian2/synapses/synapses.py]: I wouldn't worry about this now. It's an internal function (made very clear by starting the name with an underscore), so we can always rename it or change the docstring if the function becomes more general.

File: brian2/synapses/synapses.py (hunk 3 above)
Comment (dsprenkels) [inline brian2/synapses/synapses.py]: :+1:

File: brian2/tests/test_synapses.py [hunk 5]
```diff
@@ -2264,4 +2280,5 @@ def test_synapses_refractory_rand():
     test_synapse_generator_random_with_condition()
     test_synapses_refractory()
     test_synapses_refractory_rand()
+    test_synapse_generator_range_noint()
```
Comment (mstimberg) [inline brian2/tests/test_synapses.py]: Almost there: the function is called ` test_range_noint_error` and not `test_synapse_generator_range_noint()` (but I'd actually prefer the latter name) :smile:

Comment (dsprenkels) [inline brian2/tests/test_synapses.py]: Oops. Should be fixed now. :)

File: brian2/tests/test_synapses.py [hunk 6]
```diff
@@ -2190,6 +2190,22 @@ def test_synapses_refractory_rand():
     assert all(target.v[:5] > 0)
     assert_equal(target.v[5:], 0)
 
+
+@attr('codegen-independent')
+def test_synapse_generator_range_noint():
+    # arguments to `range` should only be integers (issue #781)
+    G = NeuronGroup(42, 'v: 1', threshold='False')
+    S = Synapses(G, G, 'w:1', 'v+=w')
+    msg = 'The "{}" argument of the range function was .+, but it needs to be an integer\.'
```
Comment (dsprenkels) [inline brian2/tests/test_synapses.py]: I just changed this to escape the last '.'.

Comment (mstimberg) [inline brian2/tests/test_synapses.py]: Good catch!

**Comment by mstimberg** [comment]
Ok, great, merging this now. Many thanks again!

**Pull request merged**